File: 백준/Gold/14226. 이모티콘/이모티콘.py
import sys
from collections import deque
def min_time_to_make_emoticons(s):
    visited=set()
    queue=deque()
    queue+=[(1,0,0)]
    visited.add((1,0))
    while queue:
        screen,clipboard,time=queue.popleft()
        if screen==s:
            return time;
        if (screen,screen) not in visited:
            visited.add((screen,screen))
            queue+=[(screen,screen,time+1)]
        if clipboard>0 and (screen+clipboard,clipboard) not in visited:
            visited.add((screen+clipboard,clipboard))
            queue+=[(screen+clipboard,clipboard,time+1)]
        if screen>0 and (screen-1,clipboard) not in visited:
            visited.add((screen-1,clipboard))
            queue+=[(screen-1,clipboard,time+1)]
s=int(sys.stdin.readline().rstrip())
print(min_time_to_make_emoticons(s))

File: 백준/Gold/14226. 이모티콘/이모티콘.cc
#include <deque>
#include <iostream>
#include <set>
using namespace std;
int min_time_to_make_emoticons(int s) {
    int screen, clipboard, time;
    set<set<int>> visited;
    deque<deque<int>> que;
    visited.insert({1, 0});
    que.push_back({1, 0, 0});
    while (!que.empty()) {
        screen = que.front()[0], clipboard = que.front()[1], time = que.front()[2];
        que.pop_front();
        if (screen == s)
            return time;
        if (visited.find({screen, screen}) == visited.end()) {
            visited.insert({screen, screen});
            que.push_back({screen, screen, time + 1});
        }
        if (visited.find({screen + clipboard, clipboard}) == visited.end()) {
            visited.insert({screen + clipboard, clipboard});
            que.push_back({screen + clipboard, clipboard, time + 1});
        }
        if (visited.find({screen - 1, clipboard}) == visited.end()) {
            visited.insert({screen - 1, clipboard});
            que.push_back({screen - 1, clipboard, time + 1});
        }
    }
}
int main(void) {
    int s;
    cin >> s;
    cout << min_time_to_make_emoticons(s);
    return 0;
}